commands/init.go

```diff
@@ -27,70 +41,174 @@ var initCmd = &cmds.Command{
 		cmdkit.BoolOption(DevnetUser, "when set, populates config bootstrap addrs with the dns multiaddrs of the user devnet and other user devnet specific bootstrap parameters"),
 	},
 	Run: func(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment) error {
-		repoDir := getRepoDir(req)
+		newConfig, err := getConfigFromOptions(req.Options)
+		if err != nil {
+			return err
+		}
+
+		repoDir, _ := req.Options[OptionRepoDir].(string)
 		if err := re.Emit(fmt.Sprintf("initializing filecoin node at %s\n", repoDir)); err != nil {
 			return err
 		}
+		rep, err := repo.CreateRepo(repoDir, newConfig)
 
-		genesisFile, _ := req.Options[GenesisFile].(string)
-		peerKeyFile, _ := req.Options[PeerKeyFile].(string)
-		autoSealIntervalSeconds, _ := req.Options[AutoSealIntervalSeconds].(uint)
-		devnetTest, _ := req.Options[DevnetTest].(bool)
-		devnetNightly, _ := req.Options[DevnetNightly].(bool)
-		devnetUser, _ := req.Options[DevnetUser].(bool)
-
-		var withMiner address.Address
-		if m, ok := req.Options[WithMiner].(string); ok {
-			var err error
-			withMiner, err = address.NewFromString(m)
-			if err != nil {
-				return err
-			}
+		defer func() {
+			if closeErr := rep.Close(); closeErr != nil {
+				if err == nil {
+					err = closeErr
+				} else {
+					err = errors.Wrap(err, closeErr.Error())
+				}
+			} // else err may be set and returned as normal
+		}()
+
+		genesisFileSource, _ := req.Options[GenesisFile].(string)
+		genesisFile, err := loadGenesis(req.Context, rep, genesisFileSource)
+		if err != nil {
+			return err
 		}
 
-		var defaultAddress address.Address
-		if m, ok := req.Options[DefaultAddress].(string); ok {
-			var err error
-			defaultAddress, err = address.NewFromString(m)
-			if err != nil {
-				return err
-			}
+		autoSealIntervalSeconds, _ := req.Options[AutoSealIntervalSeconds].(uint)
+		peerKeyFile, _ := req.Options[PeerKeyFile].(string)
+		initopts, err := getNodeInitOpts(autoSealIntervalSeconds, peerKeyFile)
+		if err != nil {
+			return err
 		}
 
-		return GetAPI(env).Daemon().Init(
-			req.Context,
-			api.RepoDir(repoDir),
-			api.GenesisFile(genesisFile),
-			api.PeerKeyFile(peerKeyFile),
-			api.WithMiner(withMiner),
-			api.DevnetTest(devnetTest),
-			api.DevnetNightly(devnetNightly),
-			api.DevnetUser(devnetUser),
-			api.AutoSealIntervalSeconds(autoSealIntervalSeconds),
-			api.DefaultAddress(defaultAddress),
-		)
+		// TODO: don't create the repo if this fails
```

Comment:
maybe kill this todo? or turn it into a note: if this fails we're not cleaning up repo